Implement an RFC compliant remote config client

ameske opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Replaces the old remote config client with an [RFC](https://docs.google.com/document/d/1u_G7TOr8wJX0dOM_zUDKuRJgxoJU_hVTd5SeaMucQUs/edit#heading=h.octuyiil30ph) compliant client that reflects the current requirements for products to receive remote configuration updates from an agent.

This also separates out the concept of a "remote config Repository" from the network layer so that the Go tracer can also use the "Repository" structure and only need to provide its own "networking" layer to integrate with remote configuration.

In addition, it updates the trace-agent and the security-agent to use the new API of the client - which forgoes channels as a means to provide updates and instead uses callback functions.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succint form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

This is basically a rewrite of the existing client, so the diff is unfortunately going to be very noisy and not the best way to review the PR.

The PR is split into 3 main areas:

- The definition of a "Repository" structure in pkg/remoteconfig. This structure is responsible for the main "update" loop as defined in the RFC as well as exposing a way to retrieve the current state of configuration for a given product as well as the current state of the entire Repository for making update requests.
- The networking layer in pkg/config/remote/client.go. This is just a wrapper around the "Repository" structure that translates between the domain types.
- The usage of the new client within the trace-agent and the security-agent.

It's probably easiest to review each portion separately as they are largely independent. (just not in a manner that allowed splitting into multiple PR's easily)

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Both APM Sampling (via the trace-agent) and CWS (via the security-agent) have remote config support. The best way to validate this change is to setup an agent with these products enabled as well as remote config support for those products enabled and use the "remote-config" command to verify that configurations are flowing as expected.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ameske committed:
Remove unneeded functionality

Remove most of the uptane/TUF validation logic that is no longer part of
the remote config client RFC. We'll build back up what is needed from
that.

ameske committed:
Initial RFC compliant implementation

ameske committed:
Attempt to reduce the amount of boilerplate needed to add a product

At the cost of having to manual manage some type casting, attempt to
reduce the amount of boilerplate needed to add a new product.

Adding a new product should now only consist of:
    - Adding the new product string and adding to the products list
    - Defining the config structure and a function to parse it
    - Adding that parse function to parseConfig in respository.go
    - Exposing a way to retrieve typed configs

Note that this also changes the concept of an "updateResult". The RFC
suggests providing updates to products that explicitly list the "new",
"updated", and "removed" configs...but both APM and LD don't care about
this and just want the current state of configs when updates come in. As
a result, this implementation does not do the more advanced reporting at
the moment.

ameske committed:
Make sure the metadata is stored with the config

We store it separately to make getting the repository state easier, but
we also need to store it WITH the config for downstream products to use
if they so desire.

ameske committed:
Add a way to subscribe to CWS configs

ameske committed:
Use the new client in the trace-agent and security-agent

ameske committed:
Add some missing comments

ameske committed:
Group the code needed to add a new product together

ameske committed:
Synchronize manipulation of the listener lists

ameske committed:
Remove outdated client tests

In the long term we are going to rely on system-tests to test the
client. In the short term we will be adding the functionality within the
agent. For now, clear out the client tests to clean up the draft PR
review.

ameske committed:
Don't invoke subscribers if nothing changed in an update

ameske committed:
Restore TUF checks

Even though these are not in the RFC, for the main Go client we want to
include them to prepare for that in the future. This restores the TUF
checks from the previous client.

- Handling only the roots by leveraging the go-tuf client
- Verifying the hash of every target file

ameske committed:
Update the root version dynamically now that we track them

ameske committed:
Improve ergonomics of updateRoot

After updating roots, we always need the latest root to immediately
validate the targets file, so lets return it from updateRoot.

ameske commented on pkg/security/module/module.go:
```diff
@@ -174,22 +173,14 @@ func (m *Module) Start() error {
 		if err != nil {
 			return err
 		}
+		c.RegisterCWSDDUpdate(m.processRemoteConfigsUpdate)
```
I remember you mentioning this but I lost track of it. This is easy to add! 👍🏻

ameske commented on pkg/config/remote/client.go:
```diff
@@ -64,125 +60,211 @@ func newClient(agentName string, products []data.Product, dialOpts ...grpc.DialO
+
+	c.m.Lock()
+	defer c.m.Unlock()
+	for _, listener := range c.apmListeners {
```
Ah good point, we would. I had finer grained update information like the RFC had but cut it back when I realized that none of the products actually want the "diff" and just want the current state when it changes. But we still need it internally for the moment to do exactly this, so I'll add that back.

ameske committed:
Give an immediate state update when registering callback functions

ameske committed:
Only call subscribers if the product's config changed

ameske committed:
s/repository/state for clarity

ameske committed:
Append new roots to the local store instead of overwriting

We only get the NEW roots from the Agent in each update, so we need to
append the new roots instead of overwriting them in the store.

ameske committed:
Only store updated roots if everything in the update validates

We don't want to be in a situation where we apply the root updates, but
there's an issue with something else in the update that causes our state
to be partially updated.

To accomplish this we use a temporary root TUF client that is a clone of
the current one. We then apply the root update to the temporary client,
and at the end if everything checks out, we make the temporary client
the new official client.

ameske committed:
Add helper functions to assist with building tests

ameske committed:
Add tests around proper creation of a repository.

ameske committed:
Add helper code for testing updates. Test basic updates.

ameske committed:
Validate the repository's current state in basic update tests

ameske committed:
More test updates

- Add tests around adding new roots, both malformed and correct. (Fix
  issue with testing util setup of roots as well)
- Add more tests around updating and manipulating files
- Add tests around malformed target files

ameske committed:
Add test involving update with two products

arbll commented on pkg/config/remote/client.go:
```diff
@@ -15,46 +16,41 @@ import (
-func newClient(agentName string, products []data.Product, dialOpts ...grpc.DialOption) (*Client, error) {
+// NewClient creates a new client
+func NewClient(agentName string, agentVersion string, products []data.Product, pollInterval time.Duration) (*Client, error) {
 	token, err := security.FetchAuthToken()
```
Note: I moved this to each poll query on main to mitigate the core-agent vs trace-agent startup race

ameske replied:
👍🏻  Will move it in this branch as well.

arbll commented on pkg/remoteconfig/repository.go:
```diff
@@ -0,0 +1,343 @@
+	//
+	// Note: This goes further than the RFC requires and validates the TUF targets metadata's signatures.
+	// This is NOT required for most clients per the RFC.
+	updatedTargets, err := unmarshalTargets(latestRoot, update.TUFTargets)
```
wouldn't this make more sense as a `validateTargets` method in the tuf partial client ?

ameske replied:
Ah yes, it would - which would remove the awkward need to have the latest root after a root update in the main update method.

ameske committed:
Move target validation logic to tufRootsClient

The tufRootsClient holds the root files needed to validate the targets
file, so it really belongs there.

ameske committed:
Incorporate fix for security token race condition

ameske committed:
Correct error with processing root updates

We need to validate the update payload with the temporary root client
and not the stored one.

ameske committed:
Update security-agent's NewClient call for changed API
